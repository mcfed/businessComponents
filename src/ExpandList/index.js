import React from 'react';
import {List, Button} from 'antd';
import PropTypes from 'prop-types';

import fetch from 'cross-fetch';

class ExpandList extends React.Component {
  state = {
    initLoading: true,
    loading: false,
    current: 1,
    data: [],
    list: []
  };

  componentDidMount() {
    this.getInitData();
  }

  componentWillReceiveProps() {
    this.getInitData();
  }

  getInitData() {
    const {pageSize} = this.props;
    let {current, initLoading} = this.state;
    this.getData(res => {
      if (current * pageSize < res.length) {
        initLoading = false;
      }
      this.setState({
        initLoading,
        data: res.slice(0, current * pageSize),
        current: 1,
        list: res
      });
    });
  }

  getData = callback => {
    const {fetchListUrl} = this.props;
    fetch(fetchListUrl, {
      method: 'GET'
    })
      .then(json => {
        return json.json();
      })
      .then(result => {
        if (result.code === 0) {
          callback(result.data.items);
        }
      });
  };

  onLoadMore = () => {
    const {pageSize} = this.props;
    let {current, list, initLoading} = this.state;
    current++;
    if (current * pageSize >= list.length) {
      initLoading = true;
    }
    this.setState(
      {
        data: list.slice(0, current * pageSize),
        current,
        initLoading
      },
      () => {
        window.dispatchEvent(new Event('resize'));
      }
    );
  };

  render() {
    const {header, renderItems} = this.props;
    const {loading, data, initLoading} = this.state;
    const loadMore =
      !initLoading && !loading ? (
        <div style={{textAlign: 'center', margin: 8}}>
          <a onClick={this.onLoadMore}>加载更多</a>
        </div>
      ) : null;

    return (
      <List
        className='demo-loadmore-list'
        loading={loading}
        itemLayout='horizontal'
        loadMore={loadMore}
        header={header}
        size='large'
        dataSource={data}
        renderItem={item => renderItems(item)}
      />
    );
  }
}

ExpandList.propTypes = {
  /**
   * header 头信息，接收string或者reactDom，默认为空
   */
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  /**
   * pageSize 分页条数，默认为10
   */
  pageSize: PropTypes.number,
  /**
   * fetchListUrl 请求的url 必填
   */
  fetchListUrl: PropTypes.string.isRequired,
  /**
   * renderItems 渲染的items子项
   */
  renderItems: PropTypes.func.isRequired
};
ExpandList.defaultProps = {
  header: '',
  pageSize: 10,
  fetchListUrl: '',
  renderItems: function() {}
};

export default ExpandList;
