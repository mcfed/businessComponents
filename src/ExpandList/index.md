# ExpandList 组件

> 封装的带有加载更多的 List 组件，支持本地分页，支持自定义 header，支持设置分页条数，支持点击加载更多，支持自定义设置 list 内容

## Usage

```

  renderItems = res => {
    return <div>{res.dbserverIp}</div>
  }

  <ExpandList header="IP Test" renderItems = {this.renderItems} pageSize="3" fetchListUrl = {`/capaa/dbfirewall/v2/sqlWhiteListSetting?id=${record.dbserverId}`} />

```

## Option

| 参数         | 说明                       | 类型             |
| ------------ | -------------------------- | ---------------- |
| header       | 需要显示到 header 的文本   | string/reactNode | 必填 |
| renderItems  | 支持自定义 list 的项目内容 | String/reactNode | 必填 |
| pageSize     | 分页条数                   | number           |  | 默认 10 |
| fetchListUrl | 需要执行的后端数据的请求   |  string          | 必填 |
