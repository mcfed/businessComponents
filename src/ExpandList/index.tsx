import React from 'react';
import {List, Button} from 'antd';
import fetch from 'cross-fetch';

interface ExpandListProps {
  header?: React.ReactNode | string;
  pageSize?: number;
  fetchListUrl: string;
  renderItems: (item: any) => React.ReactNode;
}

interface initializeState {
  /**
   * 用来判断是否要显示加载更多按钮
   */
  initLoading: boolean;
  /**
   * 当卡片内容还在加载中时，可以用 loading 展示一个占位
   */
  loading: boolean;
  /**
   * 列表当前所在页数
   */
  current: number;
  /**
   * 列表数据源
   */
  data: any;
  /**
   * 获取初始化时列表的全部数据
   */
  list: any;
  /**
   * pageSize 分页条数，默认为10
   */
  pageSize: number;
}

class ExpandList extends React.Component<ExpandListProps, initializeState> {
  static defaultProps = {
    header: '',
    pageSize: 10,
    fetchListUrl: '',
    renderItems: function(item: any) {}
  };

  constructor(props: ExpandListProps) {
    super(props);
    let {pageSize} = this.props;
    this.state = {
      initLoading: true,
      loading: false,
      current: 1,
      data: [],
      list: [],
      pageSize: pageSize || 1
    };
  }

  componentDidMount() {
    this.getInitData();
  }

  componentWillReceiveProps() {
    this.getInitData();
  }

  getInitData() {
    let {pageSize} = this.props;
    if (typeof pageSize === 'undefined') {
      pageSize || 10;
    }
    let {current, initLoading} = this.state;
    let totalnow = !pageSize ? current * 10 : current * pageSize;
    this.getData((res: any) => {
      res = res || [];
      if (totalnow < res.length) {
        initLoading = false;
      }
      this.setState({
        initLoading,
        data: res.slice(0, totalnow),
        current: 1,
        list: res
      });
    });
  }

  getData = (callback: any) => {
    const {fetchListUrl} = this.props;
    fetch(fetchListUrl, {
      method: 'GET'
    })
      .then((json: any) => {
        return json.json();
      })
      .then((result: any) => {
        if (result.code === 0) {
          callback(result.data.items);
        }
      });
  };

  onLoadMore = () => {
    const {pageSize} = this.props;
    let {current, list, initLoading} = this.state;
    current++;
    let totalnow = !pageSize ? current * 10 : current * pageSize;
    if (totalnow >= list.length) {
      initLoading = true;
    }
    this.setState(
      {
        data: list.slice(0, totalnow),
        // current,
        initLoading
      },
      () => {
        window.dispatchEvent(new Event('resize'));
      }
    );
  };

  renderItems = (item: any) => {
    const {renderItems} = this.props;
    if (!renderItems) return null;
    return renderItems(item);
  };

  render() {
    const {header, renderItems} = this.props;
    const {loading, data, initLoading} = this.state;
    const loadMore =
      !initLoading && !loading ? (
        <div style={{textAlign: 'center', margin: 8}}>
          <a onClick={this.onLoadMore}>加载更多</a>
        </div>
      ) : null;

    return (
      <List
        className='demo-loadmore-list'
        loading={loading}
        itemLayout='horizontal'
        loadMore={loadMore}
        header={header}
        size='large'
        dataSource={data}
        renderItem={(item: any) => this.renderItems(item)}
      />
    );
  }
}

export default ExpandList;
