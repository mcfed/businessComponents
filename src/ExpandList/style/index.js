import 'antd/es/list/style';
