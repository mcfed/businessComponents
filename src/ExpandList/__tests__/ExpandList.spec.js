import {shallow, mount, render} from 'enzyme';
import React from 'react';

import ExpandList from '../index.tsx';

const setup = props => {
  const renderItems = item => <div>{item}</div>;
  const wrapper = shallow(
    <ExpandList {...props} fetchListUrl='test' renderItems={renderItems} />
  );
  return {
    wrapper,
    props
  };
};

describe('快照测试', () => {
  it('ExpandList 快照测试', () => {
    const {wrapper} = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('ExpandList render 是否正确渲染', () => {
  it('render', () => {
    const {wrapper} = setup();
    expect(wrapper.find('List').exists()).toBe(true);
  });

  it('默认参数是否正确渲染', () => {
    const {wrapper} = setup();
    expect(wrapper.find('List').prop('header')).toEqual('');
  });

  it('传入header', () => {
    const {wrapper} = setup({
      header: 'Test'
    });
    expect(wrapper.find('List').prop('header')).toEqual('Test');
  });
});

describe('方法测试全覆盖', () => {
  it('getInitData 是否正确调用', () => {
    const {wrapper} = setup({
      pageSize: 2
    });
    wrapper.instance().state.current = 1;
    wrapper.instance().state.initLoading = true;
    wrapper.instance().getData = jest.fn();
    wrapper.instance().getInitData();
    expect(wrapper.instance().state.initLoading).toBe(true);
  });

  it.skip('getData 测试回调函数应该能够被调用', () => {
    let mockFn = jest.fn();
    const resp = {data: {data: {}}};
    axios.get.mockResolvedValue(resp);
    const {wrapper} = setup();
    wrapper
      .instance()
      .getData(mockFn)
      .then(data => expect(mockFn(resp)).toBeCalled());
  });

  it('onLoadMore 中 current 是否正确调用', () => {
    const {wrapper} = setup({
      pageSize: 2
    });
    wrapper.instance().state.current = 1;
    wrapper.instance().onLoadMore();
    expect(wrapper.instance().state.current).toEqual(1);
  });

  it('onLoadMore 中 data 是否正确调用', () => {
    const {wrapper} = setup({
      pageSize: 2
    });
    const list = [1, 2, 3, 4, 5];
    const data = [1, 2, 3, 4];
    wrapper.instance().state.list = list;
    wrapper.instance().state.current = 1;
    wrapper.instance().onLoadMore();
    expect(wrapper.instance().state.data).toEqual(data);
  });

  it('onLoadMore 中分页未达到上限时 initLoading 是否正确调用', () => {
    const {wrapper} = setup({
      pageSize: 2
    });
    const list = [1, 2, 3, 4, 5];
    wrapper.instance().state.list = list;
    wrapper.instance().state.current = 1;
    wrapper.instance().state.initLoading = false;
    wrapper.instance().onLoadMore();
    expect(wrapper.instance().state.initLoading).toBe(false);
  });

  it('onLoadMore 中分页达到上限时 initLoading 是否正确调用', () => {
    const {wrapper} = setup({
      pageSize: 2
    });
    const list = [1, 2, 3, 4, 5];
    wrapper.instance().state.list = list;
    wrapper.instance().state.current = 2;
    wrapper.instance().state.initLoading = false;
    wrapper.instance().onLoadMore();
    expect(wrapper.instance().state.initLoading).toBe(true);
  });
});
