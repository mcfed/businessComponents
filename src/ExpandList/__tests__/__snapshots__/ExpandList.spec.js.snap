// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 ExpandList 快照测试 1`] = `
<List
  bordered={false}
  className="demo-loadmore-list"
  dataSource={Array []}
  header=""
  itemLayout="horizontal"
  loadMore={null}
  loading={false}
  pagination={false}
  renderItem={[Function]}
  size="large"
  split={true}
/>
`;
